### PR TITLE
[Fix]Fix is_module_wrapper

### PR DESCRIPTION
**Pull request by HIT-cwh**
## Motivation
In order to avoid scope collision with other codebases, DistributedDataParallelWrapper in mmrazor, mmpose, mmedit or some other codebases need to be registered to a separate registry whose parent is the MODULE_WRAPPERS registry in mmcv. Ref to [this pr](https://github.com/open-mmlab/mmpose/pull/1204). When checking if a module is a module wrapper, we need to consider these child MODULE_WRAPPERS.

## Modification
A Depth-First-Search is needed in ``is_module_wrapper``.

### CODE REVIEW

File: mmcv/parallel/utils.py [hunk 1]
```diff
@@ -8,13 +8,23 @@ def is_module_wrapper(module):
...
     Args:
         module (nn.Module): The module to be checked.
 
     Returns:
         bool: True if the input module is a module wrapper.
     """
-    module_wrappers = tuple(MODULE_WRAPPERS.module_dict.values())
-    return isinstance(module, module_wrappers)
+
+    def dfs(MODULE_WRAPPER):
```
Comment (zhouzaida) [inline mmcv/parallel/utils.py:21]:
```suggestion
    def is_module_in_wrapper(module_wrapper):
```

File: tests/test_parallel.py [hunk 2]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
     module_wraper = ModuleWrapper(model)
```
Comment (zhouzaida) [inline tests/test_parallel.py:73]: wrapped_module

File: tests/test_parallel.py [hunk 3]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
...
+    @MODULE_WRAPPERS_MMPOSE.register_module()
+    class ModuleWrapperPose(object):
+
+        def __init__(self, module):
+            self.module = module
+
+        def forward(self, *args, **kwargs):
+            return self.module(*args, **kwargs)
+
+    module_wraper = ModuleWrapperRazor(model)
```
Comment (zhouzaida) [inline tests/test_parallel.py:100]:
```suggestion
    wrapped_module = ModuleWrapperRazor(model)
```

File: tests/test_parallel.py [hunk 4]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
...
+    assert is_module_wrapper(module_wraper)
+
+    module_wraper = ModuleWrapperPose(model)
```
Comment (zhouzaida) [inline tests/test_parallel.py:103]:
```suggestion
    wrapped_module = ModuleWrapperPose(model)
```

File: tests/test_parallel.py [hunk 5]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
     module_wraper = ModuleWrapper(model)
     assert is_module_wrapper(module_wraper)
 
+    # test module wrapper registry in downstream repo
+    MODULE_WRAPPERS_MMRAZOR = Registry(
```
Comment (zhouzaida) [inline tests/test_parallel.py:77]:
```suggestion
    MMRAZOR_MODULE_WRAPPERS = Registry(
```

File: tests/test_parallel.py [hunk 6]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
...
+        'mmrazor module wrapper', parent=MODULE_WRAPPERS, scope='mmrazor')
+    MODULE_WRAPPERS_MMPOSE = Registry(
```
Comment (zhouzaida) [inline tests/test_parallel.py:79]:
```suggestion
    MMPOSE_MODULE_WRAPPERS = Registry(
```

File: mmcv/parallel/utils.py [hunk 7]
```diff
@@ -8,13 +8,23 @@ def is_module_wrapper(module):
     The following 3 modules in MMCV (and their subclasses) are regarded as
     module wrappers: DataParallel, DistributedDataParallel,
     MMDistributedDataParallel (the deprecated version). You may add you own
-    module wrapper by registering it to mmcv.parallel.MODULE_WRAPPERS.
+    module wrapper by registering it to mmcv.parallel.MODULE_WRAPPERS or
+    its child registry.
```
Comment (zhouzaida) [inline mmcv/parallel/utils.py:12]:
```suggestion
    its children registries.
```

File: mmcv/parallel/utils.py [hunk 8]
```diff
@@ -8,13 +8,23 @@ def is_module_wrapper(module):
...
         bool: True if the input module is a module wrapper.
     """
-    module_wrappers = tuple(MODULE_WRAPPERS.module_dict.values())
-    return isinstance(module, module_wrappers)
+
+    def dfs(MODULE_WRAPPER):
+        module_wrappers = tuple(MODULE_WRAPPER.module_dict.values())
+        if isinstance(module, module_wrappers):
+            return True
+        for CHILD_MODULE_WRAPPER in MODULE_WRAPPER.children.values():
```
Comment (zhouzaida) [inline mmcv/parallel/utils.py:25]:
```suggestion
        for child in MODULE_WRAPPER.children.values():
```

File: mmcv/parallel/utils.py [hunk 9]
```diff
@@ -8,13 +8,23 @@ def is_module_wrapper(module):
...
     Args:
         module (nn.Module): The module to be checked.
 
     Returns:
         bool: True if the input module is a module wrapper.
     """
-    module_wrappers = tuple(MODULE_WRAPPERS.module_dict.values())
-    return isinstance(module, module_wrappers)
+
+    def is_module_in_wrapper(module_wrapper):
```
Comment (zhouzaida) [inline mmcv/parallel/utils.py:21]:
```suggestion
    def is_module_in_wrapper(module, module_wrapper):
```

File: tests/test_parallel.py [hunk 10]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
...
     assert is_module_wrapper(module_wraper)
 
+    # test module wrapper registry in downstream repo
+    MMRAZOR_MODULE_WRAPPERS = Registry(
+        'mmrazor module wrapper', parent=MODULE_WRAPPERS, scope='mmrazor')
+    MMPOSE_MODULE_WRAPPERS = Registry(
+        'mmpose module wrapper', parent=MODULE_WRAPPERS, scope='mmpose')
+
+    @MMRAZOR_MODULE_WRAPPERS.register_module()
+    class ModuleWrapperRazor(object):
```
Comment (zhouzaida) [inline tests/test_parallel.py:83]:
```suggestion
    class ModuleWrapperInRazor(object):
```

File: tests/test_parallel.py [hunk 11]
```diff
@@ -72,6 +73,36 @@ def forward(self, *args, **kwargs):
...
+
+        def __init__(self, module):
+            self.module = module
+
+        def forward(self, *args, **kwargs):
+            return self.module(*args, **kwargs)
+
+    @MMPOSE_MODULE_WRAPPERS.register_module()
+    class ModuleWrapperPose(object):
```
Comment (zhouzaida) [inline tests/test_parallel.py:92]:
```suggestion
    class ModuleWrapperInPose(object):
```

**Review by zhouzaida — APPROVED** [review]

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**